File: backend/schemas.py
"""Pydantic models for API request/response validation."""
from datetime import datetime
from enum import Enum
from typing import Optional, List, Tuple, Dict, Any
from pydantic import BaseModel, field_validator


# ==================== Enums ====================

class DetectionStatus(str, Enum):
    recognized = "recognized"
    flagged = "flagged"
    unknown = "unknown"


class ProfileRole(str, Enum):
    employee = "employee"
    vip = "vip"
    visitor = "visitor"
    blacklist = "blacklist"
    watchlist = "watchlist"


class EmbeddingStatus(str, Enum):
    indexed = "indexed"
    pending = "pending"
    stale = "stale"
    missing = "missing"


class Gender(str, Enum):
    male = "male"
    female = "female"
    unknown = "unknown"


class EventPriority(str, Enum):
    critical = "critical"
    high = "high"
    normal = "normal"
    low = "low"


class CameraStatus(str, Enum):
    online = "online"
    degraded = "degraded"
    offline = "offline"


# ==================== Camera ====================

class CameraBase(BaseModel):
    id: str
    name: str
    zone: Optional[str] = None
    ip_address: Optional[str] = None
    rtsp_url: Optional[str] = None


class CameraResponse(CameraBase):
    status: str
    ping_ms: int
    frame_latency_ms: int
    fps: float
    gpu_load: float
    cpu_load: float
    last_heartbeat: datetime
    detections_today: int
    
    class Config:
        from_attributes = True


# ==================== Profile ====================

class ProfileBase(BaseModel):
    name: str
    role: Optional[str] = ProfileRole.visitor
    department: Optional[str] = None


class ProfileResponse(ProfileBase):
    id: str
    embedding_status: str
    embedding_count: int
    enrolled_at: datetime
    last_seen: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class ProfileUpdateRequest(BaseModel):
    name: Optional[str] = None
    role: Optional[str] = None
    department: Optional[str] = None


class UnregisteredSubjectResponse(BaseModel):
    id: str
    display_name: str
    capture_count: int
    first_seen: datetime
    last_seen: datetime
    cameras: List[str]
    best_confidence: float
    representative_fingerprint: str
    vector_dimension: int
    event_ids: List[str]
    status: str


class UnregisteredSubjectRenameRequest(BaseModel):
    display_name: str


class UnregisteredSubjectRegisterRequest(BaseModel):
    name: str
    role: str = "visitor"
    department: Optional[str] = None


class UnregisteredSubjectAssignRequest(BaseModel):
    profile_id: str


class UnregisteredSubjectMergeRequest(BaseModel):
    source_subject_id: str


# ==================== Detection / FaceLog ====================

class DetectionBase(BaseModel):
    camera_id: str
    profile_id: Optional[str] = None
    timestamp: datetime
    status: str = DetectionStatus.unknown
    confidence: float = 0.0
    liveness_score: float = 0.0
    age: Optional[int] = None
    gender: Optional[Gender] = Gender.unknown
    wearing_mask: Optional[bool] = False
    wearing_glasses: Optional[bool] = False
    bbox: List[int]  # [x1, y1, x2, y2]
    priority: Optional[EventPriority] = EventPriority.normal

    @field_validator('bbox', mode='before')
    @classmethod
    def parse_bbox(cls, v):
        if isinstance(v, str):
            import json
            try:
                return json.loads(v)
            except ValueError:
                return []
        return v


class DetectionCreateRequest(BaseModel):
    camera_id: str
    identity: str
    confidence: float
    bbox: List[int]
    timestamp: datetime
    age: Optional[int] = None
    gender: Optional[str] = None
    event_id: str  # Required idempotency key (deterministic SHA-256 hash from edge node)
    device_id: Optional[str] = None
    sequence_number: Optional[int] = None
    priority: Optional[EventPriority] = EventPriority.normal
    config_version: Optional[int] = 1
    
    # AI Model and Configuration Version Tracking
    detection_model_version: Optional[str] = "scrfd_500m_bnkps_v1"
    embedding_model_version: Optional[str] = "w600k_mbf_v1"
    gallery_version: Optional[int] = 1
    threshold_version: Optional[int] = 1
    camera_config_version: Optional[int] = 1
    algorithm_version: Optional[str] = "temporal_fusion_v2"
    version_bundle_hash: Optional[str] = None
    embedding: Optional[List[float]] = None
    provenance: Optional[Dict[str, Any]] = None

class DetectionBatchRequest(BaseModel):
    detections: List[DetectionCreateRequest]

class SequenceSyncInfo(BaseModel):
    device_id: str
    camera_id: str
    last_acknowledged_sequence: int
    is_duplicate: bool
    is_out_of_order: bool
    is_gap_detected: bool

class DetectionResponse(DetectionBase):
    id: str
    event_id: Optional[str] = None
    device_id: Optional[str] = None
    sequence_number: Optional[int] = None
    config_version: Optional[int] = 1
    
    detection_model_version: Optional[str] = None
    embedding_model_version: Optional[str] = None
    gallery_version: Optional[int] = None
    threshold_version: Optional[int] = None
    camera_config_version: Optional[int] = None
    algorithm_version: Optional[str] = None
    version_bundle_hash: Optional[str] = None
    
    camera: Optional[CameraResponse] = None
    profile: Optional[ProfileResponse] = None
    sync_info: Optional[SequenceSyncInfo] = None
    inserted: bool = True  # True = newly created; False = duplicate, existing row returned
    
    class Config:
        from_attributes = True


# ==================== Camera Configuration ====================

class CameraConfigBase(BaseModel):
    detection_threshold: float = 0.50
    recognition_threshold: float = 0.35
    quality_thresholds: Optional[dict] = None
    sampling_rate: int = 1
    temporal_window: float = 3.0
    notes: Optional[str] = None

    @field_validator('quality_thresholds', mode='before')
    @classmethod
    def parse_quality_thresholds(cls, v):
        if isinstance(v, str):
            import json
            try:
                return json.loads(v)
            except Exception:
                return None
        return v


class CameraConfigUpdateRequest(CameraConfigBase):
    pass


class CameraConfigRollbackRequest(BaseModel):
    target_version: int
    notes: Optional[str] = None


class CameraConfigResponse(CameraConfigBase):
    id: str
    camera_id: str
    version: int
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class CameraConfigHistoryResponse(BaseModel):
    camera_id: str
    active_version: int
    history: List[CameraConfigResponse]


class FaceLogResponse(BaseModel):
    """Detailed face detection log matching frontend FaceLog type."""
    id: str
    camera_id: str
    camera_name: str
    timestamp: str
    status: str
    confidence: float
    liveness_score: float
    profile_id: Optional[str]
    profile_name: Optional[str]
    role: Optional[str]
    age: int
    gender: str
    wearing_mask: bool
    wearing_glasses: bool
    snapshot_tone: str  # Color token for avatar


# ==================== Alerts ====================

class AlertResponse(BaseModel):
    id: str
    log_id: str
    camera_id: str
    camera_name: str
    timestamp: str
    severity: str  # critical, high, medium
    reason: str
    profile_id: Optional[str]
    profile_name: str
    role: str
    confidence: float
    acknowledged: bool
    snapshot_tone: str
    
    class Config:
        from_attributes = True


# ==================== KPIs ====================

class SystemKpisResponse(BaseModel):
    """System-wide key performance indicators."""
    total_detections: int
    unique_individuals: int
    total_profiles: int
    cameras_online: int
    critical_alerts: int
    recognitions_today: int
    unknowns_today: int
    average_confidence: float


# ==================== Thresholds ====================

class ModelThreshold(BaseModel):
    name: str
    value: float
    description: Optional[str] = None


class ModelThresholdsResponse(BaseModel):
    """Configurable model thresholds for detection/recognition."""
    similarity_confidence: float
    liveness_threshold: float
    age_variance: float
    
    class Config:
        from_attributes = True


# ==================== Forensic Search ====================

class ForensicMatchResponse(BaseModel):
    """Result of forensic face search against gallery."""
    profile_id: str
    profile_name: str
    role: Optional[str] = None
    match_score: float
    embeddings_matched: int
    last_seen: Optional[datetime]
    camera_name: Optional[str] = None
    avatarTone: str = "sky"


class ForensicVectorSearchRequest(BaseModel):
    """Forensic search request with probe embedding generated by an edge node."""
    embedding: List[float]
    threshold: float = 0.60
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    camera_ids: Optional[List[str]] = None
    gender: Optional[str] = None
    age_min: Optional[int] = None
    age_max: Optional[int] = None
    wearing_mask: Optional[bool] = None
    wearing_glasses: Optional[bool] = None


# ==================== Attendance ====================

class AttendanceRecordResponse(BaseModel):
    profile_id: str
    profile_name: str
    first_seen: datetime
    last_seen: datetime
    detection_count: int
    status: str  # present, absent, late


# ==================== Duplicate Detection ====================

class DuplicateCandidateResponse(BaseModel):
    id: str
    profileAId: str
    profileAName: str
    profileARole: str
    profileAAvatarTone: str
    profileBId: str
    profileBName: str
    profileBRole: str
    profileBAvatarTone: str
    cosineSimilarity: float
    sharedSightings: int


# ==================== Trajectory ====================

class CameraSyncMetadata(BaseModel):
    camera_id: str
    highest_local_sequence: int
    lowest_pending_sequence: Optional[int] = None
    last_completed_sequence: Optional[int] = None

class SyncReconciliationRequest(BaseModel):
    device_id: str
    cameras: List[CameraSyncMetadata]

class CameraSyncRanges(BaseModel):
    camera_id: str
    missing_ranges: List[Tuple[int, int]]

class SyncReconciliationResponse(BaseModel):
    reconciled_cameras: List[CameraSyncRanges]


class TrajectoryNodeResponse(BaseModel):
    cameraId: str
    cameraName: str
    zone: Optional[str]
    timestamp: datetime
    confidence: float
    snapshotTone: str


class SubjectTrajectoryResponse(BaseModel):
    profileId: str
    profileName: str
    role: str
    path: List[TrajectoryNodeResponse]


class MovementEdgeResponse(BaseModel):
    fromCameraId: str
    fromCameraName: str
    toCameraId: str
    toCameraName: str
    count: int
    lastSeen: datetime
    averageTravelSeconds: float


# ==================== Camera Topology & Cross-Camera Continuity ====================

class CameraEdgeCreateRequest(BaseModel):
    from_camera_id: str
    to_camera_id: str
    min_travel_seconds: float = 2.0
    max_travel_seconds: float = 120.0
    typical_travel_seconds: float = 15.0
    distance_meters: float = 10.0
    transition_probability: float = 1.0
    bidirectional: bool = False


class CameraEdgeResponse(BaseModel):
    from_camera_id: str
    to_camera_id: str
    min_travel_seconds: float
    max_travel_seconds: float
    typical_travel_seconds: float
    distance_meters: float
    transition_probability: float
    bidirectional: bool


class CameraNodeResponse(BaseModel):
    camera_id: str
    name: str
    zone: Optional[str] = ""
    location_description: Optional[str] = ""
    coordinates: Optional[Tuple[float, float]] = None


class CameraTopologyResponse(BaseModel):
    nodes: Dict[str, CameraNodeResponse]
    edges: List[CameraEdgeResponse]


class CrossCameraEvaluationRequest(BaseModel):
    from_camera_id: str
    to_camera_id: str
    elapsed_seconds: float
    embedding_similarity: float


class CrossCameraEvaluationResponse(BaseModel):
    from_camera: str
    to_camera: str
    elapsed_seconds: float
    expected_travel_range: List[float]
    temporal_score: float
    embedding_similarity: float
    topology_edge_exists: bool
    is_teleportation: bool
    is_expired: bool
    classification: str
    explanation: str


# ==================== Cloud Vector Search & Gallery Versioning ====================

class VectorSearchRequest(BaseModel):
    embedding: List[float]
    top_k: int = 1
    threshold: float = 0.50
    embedding_model_version: str = "w600k_mbf_v1"


class VectorSearchMatch(BaseModel):
    identity: str
    score: float
    profile_id: str
    model_version: str = "w600k_mbf_v1"


class VectorSearchResponse(BaseModel):
    matches: List[VectorSearchMatch]
    search_latency_ms: float
    queried_model_version: str = "w600k_mbf_v1"


class VersionBundleResponse(BaseModel):
    detection_model_version: str
    embedding_model_version: str
    gallery_version: int
    threshold_version: int
    camera_config_version: int
    algorithm_version: str
    bundle_hash: str


class GalleryResponse(BaseModel):
    version: int
    labels: List[str]
    embeddings: List[List[float]]
    profile_ids: List[str]
    synced_at: datetime


# ==================== Edge Node Health & Adaptive Runtime ====================

class NodeHealthReportRequest(BaseModel):
    device_id: str
    camera_id: Optional[str] = None
    mode: str = "NORMAL"
    metrics: Dict[str, Any]
    decisions: Optional[List[Dict[str, Any]]] = None


class NodeHealthReportResponse(BaseModel):
    status: str
    recorded_at: datetime


# ==================== Recognition Provenance Lineage ====================

class ProvenanceCandidateResponse(BaseModel):
    identity: str
    score: float
    profile_id: Optional[str] = None
    rank: int = 1


class ProvenanceStageResponse(BaseModel):
    stage_name: str
    stage_id: str
    timestamp: float
    metadata: Dict[str, Any]


class ProvenanceResponse(BaseModel):
    event_id: str
    detection_id: Optional[str] = None
    camera_id: str
    camera_config_version: int = 1
    frame_reference: str
    track_id: Optional[str] = None
    observation_references: List[str]
    detection_model_version: str
    embedding_model_version: str
    embedding_fingerprint: str
    candidate_matches: List[ProvenanceCandidateResponse]
    decision_tier: str
    selected_identity: str
    confidence: float
    decision_timestamp: datetime
    sync_event_id: Optional[str] = None
    cloud_record_id: Optional[str] = None
    provenance_chain_hash: str
    stages: List[ProvenanceStageResponse]


class ProvenanceRetentionRequest(BaseModel):
    max_retention_days: int = 30


class ProvenanceRetentionResponse(BaseModel):
    purged_records_count: int
    retained_records_count: int
    cutoff_timestamp: datetime




class MovementNetworkResponse(BaseModel):
    edges: List[MovementEdgeResponse]


# ==================== Footfall ====================

class FootfallBucketResponse(BaseModel):
    hour: str
    detections: int
    recognized: int
    unknown: int


# ==================== Demographics ====================

class DemographicSliceResponse(BaseModel):
    label: str
    value: int


# ==================== System KPIs (Full frontend shape) ====================

class SystemKpisFullResponse(BaseModel):
    connectedCameras: int
    totalCameras: int
    detectionsToday: int
    activeAlerts: int
    systemHealth: str
    gpuLoad: float
    cpuLoad: float
    avgLatencyMs: float


# ==================== Forensic Match (Full frontend shape) ====================

class ForensicMatchFullResponse(BaseModel):
    profileId: str
    profileName: str
    role: str
    cosineSimilarity: float
    lastSeen: Optional[datetime]
    cameraName: Optional[str]
    avatarTone: str


# ==================== Attendance (Full frontend shape) ====================

class AttendanceRecordFullResponse(BaseModel):
    profileId: str
    profileName: str
    role: str
    department: Optional[str]
    checkIn: Optional[datetime]
    checkOut: Optional[datetime]
    totalSightings: int
    avatarTone: str


# ==================== Alert Acknowledge ====================

class AlertAcknowledgeRequest(BaseModel):
    acknowledged: bool = True


# ==================== Profile Merge ====================

class ProfileMergeRequest(BaseModel):
    profileAId: str
    profileBId: str
    keepProfile: Optional[str] = None
    keepProfileId: Optional[str] = None
    deleteMerged: Optional[bool] = True


# ==================== Profile Create ====================

class ProfileCreateRequest(BaseModel):
    name: str
    role: Optional[str] = "visitor"
    department: Optional[str] = None
    age: Optional[int] = None
    gender: Optional[str] = "unknown"
    notes: Optional[str] = None

File: facial-recognition-dashboard/lib/schemas.ts
import { z } from 'zod';
import { AVATAR_TONES, hashString, normalizeRole } from './api/config';
import type { ProfileRole } from './types';

// Helpers
const strOrEmpty = z.string().nullish().transform(v => v || '');
const numOrZero = z.coerce.number().nullish().transform(v => v || 0);
const boolOrFalse = z.coerce.boolean().nullish().transform(v => v || false);

// Converts 0.0-1.0 to 0-100%
const confidenceTransform = z.coerce.number().nullish().transform(v => {
  if (v === undefined || v === null) return 0;
  return v > 1 ? v : v * 100;
});

// Enums
export const SystemHealthSchema = z.enum(['green', 'yellow', 'red']).catch('red');
export const DetectionStatusSchema = z.enum(['recognized', 'flagged', 'unknown']).catch('unknown');
export const ProfileRoleSchema = z.enum(['employee', 'vip', 'visitor', 'blacklist', 'watchlist']).catch('visitor');
export const EmbeddingStatusSchema = z.enum(['indexed', 'pending', 'stale', 'missing']).catch('missing');
export const GenderSchema = z.enum(['male', 'female', 'unknown']).catch('unknown');
export const CameraStatusSchema = z.enum(['online', 'degraded', 'offline']).catch('offline');
export const AgeBracketSchema = z.enum(['0-17', '18-25', '26-35', '36-50', '51-65', '65+']).catch('0-17');

// Models
export const CameraSchema = z.object({
  id: strOrEmpty,
  name: strOrEmpty,
  zone: strOrEmpty,
  ip_address: z.string().optional(),
  ipAddress: z.string().optional(),
  rtsp_url: z.string().optional(),
  rtspUrl: z.string().optional(),
  status: CameraStatusSchema,
  ping_ms: z.number().optional(),
  pingMs: z.number().optional(),
  frame_latency_ms: z.number().optional(),
  frameLatencyMs: z.number().optional(),
  fps: numOrZero,
  gpu_load: z.number().optional(),
  gpuLoad: z.number().optional(),
  cpu_load: z.number().optional(),
  cpuLoad: z.number().optional(),
  last_heartbeat: z.string().optional(),
  lastHeartbeat: z.string().optional(),
  detections_today: z.number().optional(),
  detectionsToday: z.number().optional()
}).transform(raw => ({
  id: raw.id,
  name: raw.name,
  zone: raw.zone,
  ipAddress: raw.ip_address ?? raw.ipAddress ?? '',
  rtspUrl: raw.rtsp_url ?? raw.rtspUrl ?? '',
  status: raw.status,
  pingMs: raw.ping_ms ?? raw.pingMs ?? 0,
  frameLatencyMs: raw.frame_latency_ms ?? raw.frameLatencyMs ?? 0,
  fps: raw.fps,
  gpuLoad: raw.gpu_load ?? raw.gpuLoad ?? 0,
  cpuLoad: raw.cpu_load ?? raw.cpuLoad ?? 0,
  lastHeartbeat: raw.last_heartbeat ?? raw.lastHeartbeat ?? '',
  detectionsToday: raw.detections_today ?? raw.detectionsToday ?? 0
}));

export const FaceLogSchema = z.object({
  id: strOrEmpty,
  camera_id: z.string().optional(),
  cameraId: z.string().optional(),
  camera_name: z.string().optional(),
  cameraName: z.string().optional(),
  timestamp: strOrEmpty,
  status: DetectionStatusSchema,
  confidence: confidenceTransform,
  liveness_score: z.number().optional(),
  livenessScore: z.number().optional(),
  profile_id: z.string().nullable().optional(),
  profileId: z.string().nullable().optional(),
  profile_name: z.string().nullable().optional(),
  profileName: z.string().nullable().optional(),
  role: z.string().nullable().optional(),
  age: numOrZero,
  gender: GenderSchema,
  wearing_mask: z.boolean().optional(),
  wearingMask: z.boolean().optional(),
  wearing_glasses: z.boolean().optional(),
  wearingGlasses: z.boolean().optional(),
  snapshot_tone: z.string().optional(),
  snapshotTone: z.string().optional()
}).transform(raw => ({
  id: raw.id,
  cameraId: raw.camera_id ?? raw.cameraId ?? '',
  cameraName: raw.camera_name ?? raw.cameraName ?? '',
  timestamp: raw.timestamp,
  status: raw.status,
  confidence: raw.confidence,
  livenessScore: Math.max(raw.liveness_score ?? raw.livenessScore ?? 0, 0),
  profileId: raw.profile_id ?? raw.profileId ?? null,
  profileName: raw.profile_name ?? raw.profileName ?? null,
  role: raw.role ? normalizeRole(raw.role, null as unknown as ProfileRole) : null,
  age: raw.age,
  gender: raw.gender,
  wearingMask: raw.wearing_mask ?? raw.wearingMask ?? false,
  wearingGlasses: raw.wearing_glasses ?? raw.wearingGlasses ?? false,
  snapshotTone: raw.snapshot_tone ?? raw.snapshotTone ?? AVATAR_TONES[hashString(raw.id) % AVATAR_TONES.length]
}));

export const AlertSchema = z.object({
  id: strOrEmpty,
  log_id: z.string().optional(),
  logId: z.string().optional(),
  camera_id: z.string().optional(),
  cameraId: z.string().optional(),
  camera_name: z.string().optional(),
  cameraName: z.string().optional(),
  timestamp: strOrEmpty,
  severity: z.enum(['critical', 'high', 'medium']).catch('medium'),
  reason: strOrEmpty,
  profile_id: z.string().nullable().optional(),
  profileId: z.string().nullable().optional(),
  profile_name: z.string().optional(),
  profileName: z.string().optional(),
  role: z.string().optional(),
  confidence: confidenceTransform,
  acknowledged: boolOrFalse,
  snapshot_tone: z.string().optional(),
  snapshotTone: z.string().optional()
}).transform(raw => ({
  id: raw.id,
  logId: raw.log_id ?? raw.logId ?? '',
  cameraId: raw.camera_id ?? raw.cameraId ?? '',
  cameraName: raw.camera_name ?? raw.cameraName ?? '',
  timestamp: raw.timestamp,
  severity: raw.severity,
  reason: raw.reason,
  profileId: raw.profile_id ?? raw.profileId ?? null,
  profileName: raw.profile_name ?? raw.profileName ?? '',
  role: normalizeRole(raw.role ?? ''),
  confidence: raw.confidence,
  acknowledged: raw.acknowledged,
  snapshotTone: raw.snapshot_tone ?? raw.snapshotTone ?? AVATAR_TONES[hashString(raw.id) % AVATAR_TONES.length]
}));

export const ProfileSchema = z.object({
  id: strOrEmpty,
  name: strOrEmpty,
  role: z.string().optional(),
  department: strOrEmpty,
  embedding_status: z.string().optional(),
  embeddingStatus: z.string().optional(),
  embedding_count: z.number().optional(),
  embeddingCount: z.number().optional(),
  enrolled_at: z.string().optional(),
  enrolledAt: z.string().optional(),
  last_seen: z.string().nullable().optional(),
  lastSeen: z.string().nullable().optional(),
  age: numOrZero,
  gender: GenderSchema,
  avatar_tone: z.string().optional(),
  avatarTone: z.string().optional(),
  notes: z.string().optional()
}).transform(raw => ({
  id: raw.id,
  name: raw.name,
  role: normalizeRole(raw.role ?? ''),
  department: raw.department,
  embeddingStatus: (['indexed', 'pending', 'stale', 'missing'].includes(raw.embedding_status ?? raw.embeddingStatus ?? '') 
    ? (raw.embedding_status ?? raw.embeddingStatus) 
    : 'missing') as unknown,
  embeddingCount: raw.embedding_count ?? raw.embeddingCount ?? 0,
  enrolledAt: raw.enrolled_at ?? raw.enrolledAt ?? '',
  lastSeen: raw.last_seen ?? raw.lastSeen ?? null,
  age: raw.age,
  gender: raw.gender,
  avatarTone: raw.avatar_tone ?? raw.avatarTone ?? AVATAR_TONES[hashString(raw.id) % AVATAR_TONES.length],
  notes: raw.notes
}));

export const UnregisteredSubjectSchema = z.object({
  id: strOrEmpty,
  display_name: z.string().optional(),
  displayName: z.string().optional(),
  capture_count: z.number().optional(),
  captureCount: z.number().optional(),
  first_seen: z.string().optional(),
  firstSeen: z.string().optional(),
  last_seen: z.string().optional(),
  lastSeen: z.string().optional(),
  cameras: z.array(z.string()).catch([]),
  best_confidence: z.number().optional(),
  bestConfidence: z.number().optional(),
  representative_fingerprint: z.string().optional(),
  representativeFingerprint: z.string().optional(),
  vector_dimension: z.number().optional(),
  vectorDimension: z.number().optional(),
  event_ids: z.array(z.string()).optional(),
  eventIds: z.array(z.string()).optional(),
  status: strOrEmpty
}).transform(raw => ({
  id: raw.id,
  displayName: raw.display_name ?? raw.displayName ?? '',
  captureCount: raw.capture_count ?? raw.captureCount ?? 0,
  firstSeen: raw.first_seen ?? raw.firstSeen ?? '',
  lastSeen: raw.last_seen ?? raw.lastSeen ?? '',
  cameras: raw.cameras,
  bestConfidence: raw.best_confidence ?? raw.bestConfidence ?? 0,
  representativeFingerprint: raw.representative_fingerprint ?? raw.representativeFingerprint ?? '',
  vectorDimension: raw.vector_dimension ?? raw.vectorDimension ?? 0,
  eventIds: raw.event_ids ?? raw.eventIds ?? [],
  status: raw.status
}));

export const DuplicateCandidateSchema = z.object({
  id: strOrEmpty,
  profileA: z.any(),
  profile_a: z.any(),
  profileB: z.any(),
  profile_b: z.any(),
  cosine_similarity: z.number().optional(),
  cosineSimilarity: z.number().optional(),
  shared_sightings: z.number().optional(),
  sharedSightings: z.number().optional()
}).transform(raw => {
  const pA = raw.profileA ?? raw.profile_a ?? {};
  const pB = raw.profileB ?? raw.profile_b ?? {};
  return {
    id: raw.id,
    profileA: {
      id: pA.id ?? '',
      name: pA.name ?? '',
      role: normalizeRole(pA.role ?? ''),
      avatarTone: pA.avatar_tone ?? pA.avatarTone ?? AVATAR_TONES[hashString(pA.id ?? '') % AVATAR_TONES.length],
    },
    profileB: {
      id: pB.id ?? '',
      name: pB.name ?? '',
      role: normalizeRole(pB.role ?? ''),
      avatarTone: pB.avatar_tone ?? pB.avatarTone ?? AVATAR_TONES[hashString(pB.id ?? '') % AVATAR_TONES.length],
    },
    cosineSimilarity: raw.cosine_similarity ?? raw.cosineSimilarity ?? 0,
    sharedSightings: raw.shared_sightings ?? raw.sharedSightings ?? 0
  };
});

export const TrajectoryNodeSchema = z.object({
  camera_id: z.string().optional(),
  cameraId: z.string().optional(),
  camera_name: z.string().optional(),
  cameraName: z.string().optional(),
  zone: strOrEmpty,
  timestamp: strOrEmpty,
  confidence: confidenceTransform,
  snapshot_tone: z.string().optional(),
  snapshotTone: z.string().optional()
}).transform(raw => ({
  cameraId: raw.camera_id ?? raw.cameraId ?? '',
  cameraName: raw.camera_name ?? raw.cameraName ?? '',
  zone: raw.zone,
  timestamp: raw.timestamp,
  confidence: raw.confidence,
  snapshotTone: raw.snapshot_tone ?? raw.snapshotTone ?? ''
}));

export const SubjectTrajectorySchema = z.object({
  profile_id: z.string().nullable().optional(),
  profileId: z.string().nullable().optional(),
  profile_name: z.string().optional(),
  profileName: z.string().optional(),
  role: z.string().nullable().optional(),
  path: z.array(TrajectoryNodeSchema).catch([])
}).transform(raw => ({
  profileId: raw.profile_id ?? raw.profileId ?? null,
  profileName: raw.profile_name ?? raw.profileName ?? '',
  role: raw.role ? normalizeRole(raw.role, null as unknown as ProfileRole) : null,
  path: raw.path
}));

export const MovementEdgeSchema = z.object({
  from_camera_id: z.string().optional(),
  fromCameraId: z.string().optional(),
  from_camera_name: z.string().optional(),
  fromCameraName: z.string().optional(),
  to_camera_id: z.string().optional(),
  toCameraId: z.string().optional(),
  to_camera_name: z.string().optional(),
  toCameraName: z.string().optional(),
  count: numOrZero,
  last_seen: z.string().optional(),
  lastSeen: z.string().optional(),
  average_travel_seconds: z.number().optional(),
  averageTravelSeconds: z.number().optional()
}).transform(raw => ({
  fromCameraId: raw.from_camera_id ?? raw.fromCameraId ?? '',
  fromCameraName: raw.from_camera_name ?? raw.fromCameraName ?? '',
  toCameraId: raw.to_camera_id ?? raw.toCameraId ?? '',
  toCameraName: raw.to_camera_name ?? raw.toCameraName ?? '',
  count: raw.count,
  lastSeen: raw.last_seen ?? raw.lastSeen ?? '',
  averageTravelSeconds: raw.average_travel_seconds ?? raw.averageTravelSeconds ?? 0
}));

export const MovementNetworkSchema = z.object({
  edges: z.array(MovementEdgeSchema).catch([])
});

export const ForensicMatchSchema = z.object({
  profile_id: z.string().nullable().optional(),
  profileId: z.string().nullable().optional(),
  profile_name: z.string().optional(),
  profileName: z.string().optional(),
  role: z.string().nullable().optional(),
  cosine_similarity: z.number().optional(),
  cosineSimilarity: z.number().optional(),
  last_seen: z.string().optional(),
  lastSeen: z.string().optional(),
  camera_name: z.string().optional(),
  cameraName: z.string().optional(),
  avatar_tone: z.string().optional(),
  avatarTone: z.string().optional()
}).transform(raw => ({
  profileId: raw.profile_id ?? raw.profileId ?? null,
  profileName: raw.profile_name ?? raw.profileName ?? '',
  role: raw.role ? normalizeRole(raw.role, null as unknown as ProfileRole) : null,
  cosineSimilarity: raw.cosine_similarity ?? raw.cosineSimilarity ?? 0,
  lastSeen: raw.last_seen ?? raw.lastSeen ?? '',
  cameraName: raw.camera_name ?? raw.cameraName ?? '',
  avatarTone: raw.avatar_tone ?? raw.avatarTone ?? AVATAR_TONES[hashString(raw.profile_id ?? raw.profileId ?? '') % AVATAR_TONES.length]
}));

export const AttendanceRecordSchema = z.object({
  profile_id: z.string().optional(),
  profileId: z.string().optional(),
  profile_name: z.string().optional(),
  profileName: z.string().optional(),
  role: z.string().optional(),
  department: strOrEmpty,
  check_in: z.string().optional(),
  checkIn: z.string().optional(),
  check_out: z.string().optional(),
  checkOut: z.string().optional(),
  total_sightings: z.number().optional(),
  totalSightings: z.number().optional(),
  avatar_tone: z.string().optional(),
  avatarTone: z.string().optional()
}).transform(raw => ({
  profileId: raw.profile_id ?? raw.profileId ?? '',
  profileName: raw.profile_name ?? raw.profileName ?? '',
  role: normalizeRole(raw.role ?? ''),
  department: raw.department,
  checkIn: raw.check_in ?? raw.checkIn ?? '',
  checkOut: raw.check_out ?? raw.checkOut ?? '',
  totalSightings: raw.total_sightings ?? raw.totalSightings ?? 0,
  avatarTone: raw.avatar_tone ?? raw.avatarTone ?? AVATAR_TONES[hashString(raw.profile_id ?? raw.profileId ?? '') % AVATAR_TONES.length]
}));

export const FootfallBucketSchema = z.object({
  hour: strOrEmpty,
  detections: numOrZero,
  recognized: numOrZero,
  unknown: numOrZero
});

export const DemographicSliceSchema = z.object({
  label: strOrEmpty,
  value: numOrZero
});

export const SystemKpisSchema = z.object({
  connectedCameras: numOrZero,
  totalCameras: numOrZero,
  detectionsToday: numOrZero,
  activeAlerts: numOrZero,
  systemHealth: SystemHealthSchema,
  gpuLoad: numOrZero,
  cpuLoad: numOrZero,
  avgLatencyMs: numOrZero,
  connected_cameras: z.number().optional(),
  total_cameras: z.number().optional(),
  detections_today: z.number().optional(),
  active_alerts: z.number().optional(),
  system_health: SystemHealthSchema.optional(),
  gpu_load: z.number().optional(),
  cpu_load: z.number().optional(),
  avg_latency_ms: z.number().optional()
}).transform(raw => ({
  connectedCameras: raw.connected_cameras ?? raw.connectedCameras,
  totalCameras: raw.total_cameras ?? raw.totalCameras,
  detectionsToday: raw.detections_today ?? raw.detectionsToday,
  activeAlerts: raw.active_alerts ?? raw.activeAlerts,
  systemHealth: raw.system_health ?? raw.systemHealth ?? 'red',
  gpuLoad: raw.gpu_load ?? raw.gpuLoad,
  cpuLoad: raw.cpu_load ?? raw.cpuLoad,
  avgLatencyMs: raw.avg_latency_ms ?? raw.avgLatencyMs
}));

export const ModelThresholdsSchema = z.object({
  recognitionConfidence: numOrZero,
  livenessScore: numOrZero,
  unknownFaceRetentionDays: numOrZero,
  autoAlertOnBlacklist: boolOrFalse,
  recognition_confidence: z.number().optional(),
  liveness_score: z.number().optional(),
  unknown_face_retention_days: z.number().optional(),
  auto_alert_on_blacklist: z.boolean().optional()
}).transform(raw => ({
  recognitionConfidence: raw.recognition_confidence ?? raw.recognitionConfidence,
  livenessScore: raw.liveness_score ?? raw.livenessScore,
  unknownFaceRetentionDays: raw.unknown_face_retention_days ?? raw.unknownFaceRetentionDays,
  autoAlertOnBlacklist: raw.auto_alert_on_blacklist ?? raw.autoAlertOnBlacklist
}));

export const UnknownCaptureSchema = z.object({
  id: strOrEmpty,
  camera_id: z.string().optional(),
  cameraId: z.string().optional(),
  camera_name: z.string().optional(),
  cameraName: z.string().optional(),
  timestamp: strOrEmpty,
  confidence: confidenceTransform,
  liveness_score: z.number().optional(),
  livenessScore: z.number().optional(),
  age: numOrZero,
  gender: GenderSchema,
  snapshot_tone: z.string().optional(),
  snapshotTone: z.string().optional()
}).transform(raw => ({
  id: raw.id,
  cameraId: raw.camera_id ?? raw.cameraId ?? '',
  cameraName: raw.camera_name ?? raw.cameraName ?? '',
  timestamp: raw.timestamp,
  confidence: raw.confidence,
  livenessScore: Math.max(raw.liveness_score ?? raw.livenessScore ?? 0, 0),
  age: raw.age,
  gender: raw.gender,
  snapshotTone: raw.snapshot_tone ?? raw.snapshotTone ?? AVATAR_TONES[hashString(raw.id) % AVATAR_TONES.length]
}));

